master/internal/api_trials.go

```diff
@@ -479,6 +480,45 @@ func (a *apiServer) GetExperimentTrials(
 		return nil, errors.Wrapf(err, "failed to get trials for experiment %d", req.ExperimentId)
 	}
 
+	tasks := []*taskv1.ExperimentAllocation{}
+	err = a.m.db.QueryProto(
+		"get_active_allocations",
+		&tasks,
+		0,
+		req.ExperimentId,
+	)
+	if err != nil {
+		return nil, err
+	}
+	applyIDs := make(map[int32]experimentv1.State)
+	for _, task := range tasks {
+		isRunning := false
+		isStarting := false
+		for _, allocation := range task.Allocations {
+			if allocation.State == taskv1.State_STATE_RUNNING {
+				isRunning = true
+				break
+			} else {
+				isStarting = true
+			}
+		}
+		if isRunning {
+			applyIDs[task.TrialId] = experimentv1.State_STATE_ACTIVE
+		} else if isStarting {
+			applyIDs[task.TrialId] = experimentv1.State_STATE_PENDING
+		}
+	}
+	for _, trial := range resp.Trials {
+		if trial.State == experimentv1.State_STATE_ACTIVE {
+			if setState, ok := applyIDs[trial.Id]; ok {
+				trial.State = setState
+			}
+			//  else {
+			// 	trial.State = experimentv1.State_STATE_QUEUED
+			// }
+		}
+	}
```

Comment:
I think we could rewrite all this code (and everything with a similar structure, basically) in a way that is generic to jobs (so not experiment or trial specific). Maybe be helpful to think of it like aggregating the job/task state from its job/task id, then converting job/task states to experiment, trial, etc states. Would save us a lot of duplication and help us get better about maintaining feature parity/identifying gaps in feature parity between job types to have everything be the same under the hood with some thin shims to fit specific types' requirements.

Reply:
My current solution for this is to return IsRunning and IsStarting from the SQL query into the struct... for whatever reason my initial query would return the enum's int values instead of strings.  The new version simplifies things but I ought to test it out on more complex Tasks and multi-trial Experiments.

Reply:
Updated with counts in the SQL for running allocations and all pending/running/starting allocations